File: genetic-recombination/src/main/java/genetic/recombinators/chromosome/ChromosomeRecombinator.java
package genetic.recombinators.chromosome;


import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Function;
import java.util.function.Supplier;

import genetic.api.individual.Cell;
import genetic.recombinators.Recombinator;
import javaslang.Tuple2;
import javaslang.collection.Seq;
import javaslang.collection.Traversable;
import org.springframework.beans.factory.annotation.Autowired;

/**
 * @author piotr.larysz
 */
public abstract class ChromosomeRecombinator<T1> implements Recombinator<Supplier<Seq<Cell<T1>>>, Seq<Cell<T1>>> {

    @Autowired
    protected Supplier<ThreadLocalRandom> random;

    @Autowired
    protected Function<Cell<T1>, Cell<T1>> cellCloningMapper;

    private final Supplier<Boolean> recombination = () -> random.get().nextDouble() <= 0.7;

    @Override
    public Traversable<Seq<Cell<T1>>> apply(Tuple2<? extends Supplier<Seq<Cell<T1>>>, ? extends Supplier<Seq<Cell<T1>>>> tuple) {
        return tuple.map(Supplier::get, Supplier::get).transform(this::proceed).map(seq -> seq.map(cellCloningMapper));
    }
    private Traversable<Seq<Cell<T1>>> proceed(Seq<Cell<T1>> first, Seq<Cell<T1>> second) {
        if(recombination.get()) {
            return this.recombine(first, second, random.get().nextInt(first.size()));
        }
        return this.resolveDefault(first, second);
    }

    protected abstract Traversable<Seq<Cell<T1>>> recombine(Seq<Cell<T1>> first, Seq<Cell<T1>> second, int exchangePoint);

    protected abstract Traversable<Seq<Cell<T1>>> resolveDefault(Seq<Cell<T1>> first, Seq<Cell<T1>> second);
}

File: genetic-recombinators/src/main/java/genetic/recombinators/chromosome/ChromosomeRecombinator.kt
package genetic.recombinators.chromosome

import genetic.recombinators.Recombinator
import java.util.concurrent.ThreadLocalRandom

abstract class ChromosomeRecombinator<T, U>(
    protected val random: () -> ThreadLocalRandom
) : Recombinator<Sequence<T>, Sequence<T>, U> {

  private val recombination: () -> Boolean = { random().nextDouble() <= 0.7 }

  override fun invoke(pair: Pair<Sequence<T>, Sequence<T>>, embryo: U) = // FIXME remove embryo? =
      pair.let { (first, second) ->
        if (recombination()) {
          recombine(
              first = first,
              second = second,
              exchangePoint = random().nextInt(first.count() + 1)
          )
        } else resolveDefault(first, second)
      }

  protected abstract fun recombine(
      first: Sequence<T>,
      second: Sequence<T>,
      exchangePoint: Int
  ): Sequence<Sequence<T>>

  protected abstract fun resolveDefault(
      first: Sequence<T>,
      second: Sequence<T>
  ): Sequence<Sequence<T>>
}